### ISSUE TITLE
Allow optionally turning off customized file manager

### ISSUE THREAD

**Issue by peiyuwang** [sbt/zinc#185]
#181 uses a VFS approach to protect `.class` files generation inside a transaction. This in theory should work for all files generated by javac, however, while it works for `.class` files, which is the majority case, `resource` (could be projects using annotation processor) and `.java` (An example is jmh that first generates `.java` files, then compiles them into `.class` files) files currently fail under this approach, see stack trace in [1].

The root cause is a current javac limitation, [somewhere in `JavacFileManager`](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b27/com/sun/tools/javac/file/JavacFileManager.java#638) there is a hard coded `instanceof` check, so when `FileObject` or `JavaFileObject` are wrapped into `ForwardingFileObject`/`ForwardingJavaFileObject`, instanceof check would fail because the check is on the forwarding object, not its wrapped object.

```
public boolean isSameFile(FileObject a, FileObject b) {
       nullCheck(a);
       nullCheck(b);
       if (!(a instanceof BaseFileObject))
           throw new IllegalArgumentException("Not supported: " + a);
       if (!(b instanceof BaseFileObject))
           throw new IllegalArgumentException("Not supported: " + b);
       return a.equals(b);
   }
```

Assume upstream change (I opened a request internally at Twitter) will take some time, we need some workaround to turn off the customized manager if project uses jmh for example.

A few options to introduce this option:
- Introduce a new boolean inc option. Downside this eventually will be removed, likely will touch a few layers in order to pass to `LocalJavaCompiler`
- Reuse the existing inc option `classfileManagerType`, so we only invoke VFS for `TransactionalManagerType`. Downside is `ClassFileManager` likely needs to a new method to optionally wrap an existing FileManager into `ForwardingFM`
- Pass an environment variable w/o touching any interfaces.

The third option seems the cleanest for such an use case. @eed3si9n what do you think? any other ideas?

[1] Compiling a jmh project

```
java.lang.RuntimeException: java.lang.IllegalArgumentException: Not supported: sbt.internal.inc.javac.WriteReportingJavaFileObject@1cfd1875
    at com.sun.tools.javac.main.Main.compile(Main.java:559)
    at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
    at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
    at sbt.internal.inc.javac.LocalJavaCompiler.run(LocalJava.scala:90)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler$$anonfun$compile$1.apply$mcV$sp(AnalyzingJavaCompiler.scala:74)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler$$anonfun$compile$1.apply(AnalyzingJavaCompiler.scala:72)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler$$anonfun$compile$1.apply(AnalyzingJavaCompiler.scala:72)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler.timed(AnalyzingJavaCompiler.scala:108)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler.compile(AnalyzingJavaCompiler.scala:72)
    at sbt.internal.inc.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply$mcV$sp(MixedAnalyzingCompiler.scala:58)
    at sbt.internal.inc.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply(MixedAnalyzingCompiler.scala:58)
    at sbt.internal.inc.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply(MixedAnalyzingCompiler.scala:58)
    at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:73)
    at sbt.internal.inc.MixedAnalyzingCompiler.compileJava$1(MixedAnalyzingCompiler.scala:57)
    at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:63)
    at sbt.internal.inc.IncrementalCompilerImpl$$anonfun$compileInternal$1.apply(IncrementalCompilerImpl.scala:132)
    at sbt.internal.inc.IncrementalCompilerImpl$$anonfun$compileInternal$1.apply(IncrementalCompilerImpl.scala:132)
    at sbt.internal.inc.Incremental$.doCompile(Incremental.scala:95)
    at sbt.internal.inc.Incremental$$anonfun$1$$anonfun$apply$1.apply(Incremental.scala:79)
    at sbt.internal.inc.Incremental$$anonfun$1$$anonfun$apply$1.apply(Incremental.scala:79)
    at sbt.internal.inc.IncrementalCommon.recompileClasses(IncrementalCommon.scala:70)
    at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:39)
    at sbt.internal.inc.Incremental$$anonfun$1.apply(Incremental.scala:78)
    at sbt.internal.inc.Incremental$$anonfun$1.apply(Incremental.scala:77)
    at sbt.internal.inc.Incremental$.manageClassfiles(Incremental.scala:122)
    at sbt.internal.inc.Incremental$.compile(Incremental.scala:77)
    at sbt.internal.inc.IncrementalCompile$.apply(Compile.scala:60)
    at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:132)
    at sbt.internal.inc.IncrementalCompilerImpl.incrementalCompile(IncrementalCompilerImpl.scala:87)
    at org.pantsbuild.zinc.Compiler.compile(Compiler.scala:207)
    at org.pantsbuild.zinc.Main$.run(Main.scala:97)
    at org.pantsbuild.zinc.Main$.main(Main.scala:16)
    at org.pantsbuild.zinc.Main.main(Main.scala)
Caused by: java.lang.IllegalArgumentException: Not supported: sbt.internal.inc.javac.WriteReportingJavaFileObject@1cfd1875
    at com.sun.tools.javac.file.JavacFileManager.isSameFile(JavacFileManager.java:647)
    at javax.tools.ForwardingJavaFileManager.isSameFile(ForwardingJavaFileManager.java:90)
    at com.sun.tools.javac.api.ClientCodeWrapper$WrappedJavaFileManager.isSameFile(ClientCodeWrapper.java:257)
    at com.sun.tools.javac.processing.JavacFiler.checkFileReopening(JavacFiler.java:532)
    at com.sun.tools.javac.processing.JavacFiler.getResource(JavacFiler.java:483)
    at org.openjdk.jmh.generators.annotations.APGeneratorDestinaton.getResource(APGeneratorDestinaton.java:56)
    at org.openjdk.jmh.generators.core.BenchmarkGenerator.readBenchmarkList(BenchmarkGenerator.java:194)
    at org.openjdk.jmh.generators.core.BenchmarkGenerator.complete(BenchmarkGenerator.java:135)
    at org.openjdk.jmh.generators.BenchmarkProcessor.process(BenchmarkProcessor.java:60)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$200(JavacProcessingEnvironment.java:91)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment$DiscoveredProcessors$ProcessorStateIterator.runContributingProcs(JavacProcessingEnvironment.java:627)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1033)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1198)
    at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
    at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
    at com.sun.tools.javac.main.Main.compile(Main.java:523)
    ... 32 more
[error] java.lang.IllegalArgumentException: Not supported: sbt.internal.inc.javac.WriteReportingJavaFileObject@1cfd1875
```

**Comment by eed3si9n** [comment]
So are you saying that #181 introduces a regression that breaks the build if it uses jmh and/or annotation processor? If so we need to turn it off by default.

**Comment by peiyuwang** [comment]
It's only jmh, annotation processor I discovered while testing #181 before is merged. (Different from jmh which both `.java` and `.class` are considered as `JavaFileObject`, AP uses `FileObject`, which I removed from the initial change, see 49f0a999068f4b220e79d76f05b9f2691690fd85)

Turning it off by default sounds fine. Which one of the options would you prefer?

**Comment by eed3si9n** [comment]
let's go with Boolean option on inc option so sbt users can turn it on or off from the build.

**Comment by peiyuwang** [comment]
Will do. Thanks!

**Comment by peiyuwang** [comment]
#188 is up for this issue

**Comment by peiyuwang** [comment]
Merged as d65959c3f6ef0ef0d6a800801878582a457db16c

**Comment by ggetv** [comment]
Just had the same issue and the actual option to use is:
```
--[no-]java-zinc-file-manager (default: True) 
Use zinc provided file manager to ensure transactional rollback. 
```
I used `--no-java-zinc-file-manager` and it worked for me. More details on the [Pants reference manual](https://www.pantsbuild.org/options_reference.html)
